crates/astria-sequencer/src/authority/state_ext.rs

```diff
@@ -0,0 +1,155 @@
+use anyhow::{
+    anyhow,
+    Context,
+    Result,
+};
+use async_trait::async_trait;
+use borsh::{
+    BorshDeserialize,
+    BorshSerialize,
+};
+use penumbra_storage::{
+    StateRead,
+    StateWrite,
+};
+use proto::native::sequencer::v1alpha1::{
+    Address,
+    ADDRESS_LEN,
+};
+use serde::{
+    Deserialize,
+    Serialize,
+};
+use tracing::instrument;
+
+/// Newtype wrapper to read and write an address from rocksdb.
+#[derive(BorshSerialize, BorshDeserialize, Debug)]
+struct SudoAddress([u8; ADDRESS_LEN]);
+
+/// Newtype wrapper to read and write a validator set from rocksdb.
+///
+/// Note: this is stored only in the nonconsensus state, thus is not part
+/// of the application state, so it's fine for it to be serialized/deserialized
+/// in a non-deterministic way (ie. with serde-json).
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub(crate) struct ValidatorSet(pub(crate) Vec<tendermint::validator::Update>);
+
+impl ValidatorSet {
+    pub(crate) fn apply_updates(&mut self, validator_updates: ValidatorSet) {
+        for update in validator_updates.0 {
+            if self
+                .0
+                .iter()
+                .map(|val| val.pub_key)
+                .collect::<Vec<tendermint::public_key::PublicKey>>()
```

Comment:
If I read this code correctly, this allocation is not necessary, just take the inner `find` and apply it directly, no need for the outer iter + map + collect + contains:
```rust
if let Some(validator) = self.0.iter_mut().find(|validator| validator.pub_key == update.pub_key) {
    validator.power = update.power;
} else {
    self.0.push(update);
}
```

Although I feel like using a hashmap here is probably more desirable (see the comment on the type def above)